File: backend/app/services/dashboard_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.models.ticket import Ticket


# ======================================================
# Helper
# ======================================================

def _ticket_query(db: Session, user):
    role = user["role"]

    if role == "Admin":
        return db.query(Ticket)

    if role == "Technician":
        return db.query(Ticket).filter(
            Ticket.assigned_to == user["id"]
        )

    # Employee
    return db.query(Ticket).filter(
        Ticket.created_by == user["id"]
    )


# ======================================================
# Dashboard Summary
# ======================================================

def get_dashboard_summary(db: Session, user):

    query = _ticket_query(db, user)

    return {

        "total_tickets": query.count(),

        "open_tickets": query.filter(
            Ticket.status == "Open"
        ).count(),

        "assigned_tickets": query.filter(
            Ticket.status == "Assigned"
        ).count(),

        "resolved_tickets": query.filter(
            Ticket.status == "Resolved"
        ).count(),

        "high_priority": query.filter(
            Ticket.priority == "High"
        ).count(),

        "medium_priority": query.filter(
            Ticket.priority == "Medium"
        ).count(),

        "low_priority": query.filter(
            Ticket.priority == "Low"
        ).count(),

    }


# ======================================================
# Recent Tickets
# ======================================================

def get_recent_tickets(
    db: Session,
    user,
    limit: int = 5,
):

    return (
        _ticket_query(db, user)
        .order_by(Ticket.created_at.desc())
        .limit(limit)
        .all()
    )


# ======================================================
# Ticket Trend
# ======================================================

def get_ticket_trend(db: Session, user):

    result = (
        _ticket_query(db, user)
        .with_entities(
            func.date(Ticket.created_at).label("date"),
            func.count(Ticket.id).label("tickets"),
        )
        .group_by(func.date(Ticket.created_at))
        .order_by(func.date(Ticket.created_at))
        .all()
    )

    return [

        {
            "date": str(row.date),
            "tickets": row.tickets,
        }

        for row in result

    ]

File: frontend/src/services/dashboardService.ts
import axios from "axios";

const API = "http://127.0.0.1:8000";

const getToken = () => localStorage.getItem("token");

const authHeaders = () => ({
  headers: {
    Authorization: `Bearer ${getToken()}`,
  },
});

// ======================================
// Dashboard Summary
// ======================================

export interface DashboardSummary {
  total_tickets: number;
  open_tickets: number;
  assigned_tickets: number;
  resolved_tickets: number;

  high_priority: number;
  medium_priority: number;
  low_priority: number;
}

// ======================================
// Recent Tickets
// ======================================

export interface RecentTicket {
  id: number;
  title: string;
  status: string;
  priority: string;
  assigned_to: string;
  created_at: string;
}

// ======================================
// Ticket Trend
// ======================================

export interface TicketTrend {
  date: string;
  tickets: number;
}

// ======================================
// Activity
// (Matches your backend response)
// ======================================

export interface Activity {
  message: string;
  time: string;
}

// ======================================
// API Calls
// ======================================

export async function getDashboardSummary(): Promise<DashboardSummary> {
  const response = await axios.get<DashboardSummary>(
    `${API}/dashboard/summary`,
    authHeaders()
  );

  return response.data;
}

export async function getRecentTickets(): Promise<RecentTicket[]> {
  const response = await axios.get<RecentTicket[]>(
    `${API}/dashboard/recent-tickets`,
    authHeaders()
  );

  return response.data;
}

export async function getTicketTrend(): Promise<TicketTrend[]> {
  const response = await axios.get<TicketTrend[]>(
    `${API}/dashboard/trend`,
    authHeaders()
  );

  return response.data;
}

export async function getRecentActivity(): Promise<Activity[]> {
  const response = await axios.get<Activity[]>(
    `${API}/dashboard/activity`,
    authHeaders()
  );

  return response.data;
}